### PR TITLE
DBaaS-204: create new CRD and controller for provisioning

### PR DESCRIPTION
**Pull request by xieshenzh**
Jira: https://issues.redhat.com/browse/DBAAS-204

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA or in issue number have been completed
- [ ] Unit tests added that prove the fix is effective or the feature works 
- [ ] Documentation added for the feature
- [ ] all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer

### CODE REVIEW

File: api/v1alpha1/dbaasprovider.go [hunk 1]
```diff
@@ -25,8 +25,10 @@ const (
 	// DBaaS condition types
 	DBaaSInventoryReadyType         string = "InventoryReady"
 	DBaaSInventoryProviderSyncType  string = "SpecSynced"
-	DBaaSConnectionProviderSyncType string = "ReadyForBinding"
 	DBaaSConnectionReadyType        string = "ConnectionReady"
+	DBaaSConnectionProviderSyncType string = "ReadyForBinding"
+	DBaaSInstanceReadyType          string = "InstanceReady"
+	DBaaSInstanceProviderSyncType   string = "SpecSynced"
```
Comment (jianrongzhang89) [inline api/v1alpha1/dbaasprovider.go:31]: SpecSynced => ProvisionReady

File: api/v1alpha1/dbaasprovider.go [hunk 2]
```diff
@@ -168,3 +178,67 @@ type DBaaSProviderInventory struct {
...
+type DBaaSInstanceSpec struct {
+	// A reference to the relevant DBaaSInventory CR
+	InventoryRef NamespacedName `json:"inventoryRef"`
+
+	// The name of this instance in the database service
+	Name string `json:"name,omitempty"`
+
+	// identifies the desired cloud infrastructure provider
+	// +kubebuilder:validation:Enum=aws;gcp;azure
+	Provider string `json:"provider"`
```
Comment (jianrongzhang89) [inline api/v1alpha1/dbaasprovider.go:192]: =>CloudProvider

File: api/v1alpha1/dbaasprovider.go [hunk 3]
```diff
@@ -168,3 +178,67 @@ type DBaaSProviderInventory struct {
...
+
+	// identifies the requested deployment region within the provider (e.g. us-east-1)
+	Region string `json:"region"`
```
Comment (jianrongzhang89) [inline api/v1alpha1/dbaasprovider.go:195]: =>CloudRegion

**Review by redhatHameed — COMMENTED** [review]
@xieshenzh LGTM is there anything remaining to work on this PR?

**Comment by xieshenzh** [comment]
> @xieshenzh LGTM is there anything remaining to work on this PR?

I am not sure if the design is finalized.

**Review by tchughesiv — CHANGES_REQUESTED** [review]
please rebase

File: controllers/dbaasinstance_controller.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+)
+
+// DBaaSInstanceReconciler reconciles a DBaaSInstance object
+type DBaaSInstanceReconciler struct {
+	*DBaaSReconciler
+}
+
+//+kubebuilder:rbac:groups=dbaas.redhat.com,resources=dbaasinstances,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=dbaas.redhat.com,resources=dbaasinstances/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=dbaas.redhat.com,resources=dbaasinstances/finalizers,verbs=update
```
Comment (tchughesiv) [inline controllers/dbaasinstance_controller.go:38]: i think we've been using resource wildcards for the other CRDs. I'm ok moving away from wildcards towards this but i think we should chose one way to do it. Either we define rbac for each dbaas resource type, or we use wildcards that account for all of them. thoughts?
https://github.com/RHEcosystemAppEng/dbaas-operator/blob/c4ca7029848e7fcb0191f1c0e1755b798109a996/controllers/dbaasinventory_controller.go#L36-L38

File: api/v1alpha1/dbaasprovider.go [hunk 5]
```diff
@@ -56,8 +57,23 @@ type DBaaSProviderSpec struct {
...
 	ConnectionKind string `json:"connectionKind"`
 
+	// InstanceKind is the name of the instance resource (CRD) defined by the provider for provisioning
+	InstanceKind string `json:"instanceKind"`
+
 	// CredentialFields indicates what information to collect from UX & how to display fields in a form
 	CredentialFields []CredentialField `json:"credentialFields"`
+
+	// AllowsFreeTrial indicates whether the provider provides free trials
+	AllowsFreeTrial bool `json:"allowsFreeTrial"`
```
Comment (tchughesiv) [inline api/v1alpha1/dbaasprovider.go:67]: should this be an optional bool instead and, if not set, we assume `false`?

File: api/v1alpha1/dbaasprovider.go [hunk 6]
```diff
@@ -56,8 +57,23 @@ type DBaaSProviderSpec struct {
...
+
+	// ProvisionURL URL for provisioning instances through database provider web portal
+	ProvisionURL string `json:"provisionUrl"`
```
Comment (tchughesiv) [inline api/v1alpha1/dbaasprovider.go:70]: should this be optional?

File: api/v1alpha1/dbaasprovider.go [hunk 7]
```diff
@@ -56,8 +57,23 @@ type DBaaSProviderSpec struct {
...
+
+	// ProvisionInstructions instructions on how to provision instances using provider web portal
+	ProvisionInstructions string `json:"provisionInstructions"`
```
Comment (tchughesiv) [inline api/v1alpha1/dbaasprovider.go:73]: should this be optional?

File: api/v1alpha1/dbaasprovider.go [hunk 8]
```diff
@@ -56,8 +57,23 @@ type DBaaSProviderSpec struct {
...
+
+	// InstanceParameterSpecs  indicates what parameters to collect from UX & how to display fields in a form in order to provision an instance
+	InstanceParameterSpecs []InstanceParameterSpec `json:"instanceParameterSpecs"`
```
Comment (tchughesiv) [inline api/v1alpha1/dbaasprovider.go:76]: should this be optional?

File: api/v1alpha1/dbaasprovider.go (hunk 5 above)
Comment (xieshenzh) [reply api/v1alpha1/dbaasprovider.go:67]: This is from the ADR. I think it is better to discuss in the ADR.

**Review by tchughesiv — APPROVED** [review]
lgtm

**Pull request merged**